### PR TITLE
Pytorch triple backward

### PR DESCRIPTION
**Pull request by asglover**
Adding triple backward support for higher order training to pytorch

### CODE REVIEW

File: openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py [hunk 1]
```diff
@@ -95,3 +184,98 @@ def double_backward_cpu(
...
+        in2_dgrad,
+        out_tgrad,
+        weights_tgrad,
+        in1_tgrad,
+        in2_tgrad,
+        graph,
+    ):
+        assert self.torch_op
+
+        in1_torch = torch.tensor(in1).to("cuda").requires_grad_(True)
```
Comment (vbharadwaj-bk) [inline openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py]: TODO-someday: I wonder if we can combine all of these derivative functions into one to compact this file.

**Comment by asglover** [comment]
I'm going to run the full test suite tomorrow. 
The changes are mostly about triple backwards, although my model was convinced that there was an error in the stream testing, and under closer inspection it looks like there was. 
I'll promote this from Draft to regular PR when it's ready for review.

File: tests/batch_test.py [hunk 2]
```diff
@@ -95,6 +104,113 @@ def test_tp_double_bwd(self, tp_and_problem):
...
+            "weights_double_grad",
+        ]:
+            self.check_result(result, fieldname)
+
+
+class TestTripleBackwardDirectOps:
+    @pytest.fixture(scope="class")
+    def problem(self, dtype, with_jax):
+        if with_jax:
+            pytest.skip("N/A for JAX")
```
Comment (vbharadwaj-bk) [inline tests/batch_test.py]: TODO-someday: we could expand this test to include JAX. But not in this commit.

File: tests/stream_test.py [hunk 3]
```diff
@@ -91,6 +91,65 @@ def conv_buffers(edge_index, tpp, gen):
     return (X, Y, W, edge_index[0], edge_index[1])
 
 
+@pytest.fixture
```
Comment (vbharadwaj-bk) [inline tests/stream_test.py]: Hmm. I don't think we should have any modifications to stream_test.py. Because triple_backward is a composition of existing ops that all work fine with streams, I see no reason why their composition shouldn't pass stream tests. We need to test anything that's implemented as a custom op to make sure that the stream information is lowered correctly onto the kernel, but then any composition of those operators should be ok. Let's shrink the diff here.

File: tests/conv_test.py [hunk 4]
```diff
@@ -137,6 +155,124 @@ def test_tp_double_bwd(self, conv_object, graph):
...
+            "weights_grad",
+            "output_grad",
+            "in1_double_grad",
+            "in2_double_grad",
+            "weights_double_grad",
+        ]:
+            self.check_result(result, fieldname)
+
+
+class TestTripleBackwardConvDirectOps:
```
Comment (vbharadwaj-bk) [inline tests/conv_test.py]: Remind me, what is the purpose of these DirectOps tests?

**Review by vbharadwaj-bk — CHANGES_REQUESTED** [review]
95% looks good. I think the stream_test.py modifications are redundant, since each of the custom ops have already been tested. But otherwise good.

File: openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py (hunk 1 above)
Comment (asglover) [reply openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py]: Do you just mean, turn them from .to("cuda") to (... , dtype=cuda). 

If so, I should definitely do this. It's considered good practice to. technically this does something funny which is create the tensor, then move it to the device, so the autograd actually does the opposite, moves it off device at the end. When it's supposed to just be on device I ought to use dtype=cuda. I'll clean this up. good point.

File: tests/stream_test.py (hunk 3 above)
Comment (asglover) [reply tests/stream_test.py]: The reason that any change was made is because the stream tests say (in the name) they are testing the deterministic mode but they are not (deterministic=false). This was masking some strange behavior. 

I can move this to a different PR or just ignore it. It is off topic for this PR. I'll remove it.

Comment (asglover) [reply tests/stream_test.py]: moved to a different pr

File: openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py (hunk 1 above)
Comment (asglover) [reply openequivariance/openequivariance/_torch/NPDoubleBackwardMixin.py]: fixed

File: tests/conv_test.py (hunk 4 above)
Comment (asglover) [reply tests/conv_test.py]: I removed them, they were a vestigial test from the first pass at tests, before I moved them to be part of the parameterized tests

**Review by vbharadwaj-bk — APPROVED** [review]
nice, merge when ci finishes 🏁

**Comment by asglover** [comment]
Please don't merge yet, I saw some interesting failures in the test suite that I'd like to investigate. I think they are not related to this PR, but I'd like to confirm before proceeding.

**Comment by asglover** [comment]
The failures observed during testing were related to the introduction of mace-torch in the previous PR. That pacakage required an old version of e3nn which had different w3j constants. Unrelated to this work. It will be cleaned up in a separate pr. 

This was tested and good to go
